src/Microsoft.Extensions.Configuration.AzureAppConfiguration/AzureAppConfigurationProvider.cs

```diff
@@ -742,5 +748,103 @@ private void UpdateCacheExpirationTime(KeyValueWatcher changeWatcher, bool succe
 
             changeWatcher.CacheExpires = DateTimeOffset.UtcNow.Add(cacheExpirationTime);
         }
+
+        private async Task ExecuteWithFailOverPolicyAsync(Func<ConfigurationClient, Task> funcToExecute, CancellationToken cancellationToken = default)
+        {
+            using IEnumerator<ConfigurationClient> clientEnumerator = _configClientManager.GetAvailableClients().GetEnumerator();
+
+            // Guaranteed to return at least 1 element
+            Debug.Assert(clientEnumerator.MoveNext());
+
+            while (true)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                try
+                {
+                    await funcToExecute(clientEnumerator.Current).ConfigureAwait(false);
+
+                    _configClientManager.UpdateClientStatus(clientEnumerator.Current, successful: true);
+
+                    break;
+                }
+                catch (AggregateException ae)
+                {
+                    if (!TryFailOver(ae, clientEnumerator))
+                    {
+                        throw;
+                    }
+                }
+                catch (RequestFailedException e)
+                {
+                    if (!TryFailOver(e, clientEnumerator))
+                    {
+                        throw;
+                    }
+                }
+            }
+        }
+
+        private bool TryFailOver(Exception e, IEnumerator<ConfigurationClient> clientEnumerator)
+        {
+            if (!ShouldFailoverForException(e))
+            {
+                return false;
+            }
+
+            _configClientManager.UpdateClientStatus(clientEnumerator.Current, successful: false);
+
+            if (!clientEnumerator.MoveNext())
+            {
+                return false;
+            }
+
+            return true;
+        }
+
+        private void UpdateCacheExpirationForWatchers(bool success)

```

Comment:
NIT: can we move this function in front of `UpdateCacheExpirationTime` so related functions are next to each other?